### PR TITLE
Update pantab to be compatible with HyperAPI v0.0.14567

### LINKED ISSUE #145: pantab is incompatible with the version of Tableau Hyper API installed on your system
Unsure how to proceed...

**Pantab install**
Requirement already satisfied: pantab in /usr/local/lib/python3.7/dist-packages (2.1.0)
Requirement already satisfied: tableauhyperapi in /usr/local/lib/python3.7/dist-packages (from pantab) (0.0.14567)
Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from pantab) (1.3.5)
Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from pantab) (1.21.5)
Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas->pantab) (2.8.2)
Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->pantab) (2018.9)
Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas->pantab) (1.15.0)
Requirement already satisfied: cffi!=1.14.3,<2,>=1.12.2 in /usr/local/lib/python3.7/dist-packages (from tableauhyperapi->pantab) (1.15.0)
Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi!=1.14.3,<2,>=1.12.2->tableauhyperapi->pantab) (2.21)

---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
[<ipython-input-4-a8d341fb4330>](https://jhvkagqxwjc-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220322-060047-RC00_436415511#) in <module>()
----> 1 import pantab

2 frames
[/usr/local/lib/python3.7/dist-packages/pantab/__init__.py](https://jhvkagqxwjc-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220322-060047-RC00_436415511#) in _check_compatibility(check, message)
     46 def _check_compatibility(check, message):
     47     if not check:
---> 48         raise NotImplementedError(message + "\n" + api_incompatibility_msg)
     49 
     50 

NotImplementedError: expected 'hyper_rowset_chunk_field_values' to have the signature 'struct hyper_error_t *(*)(struct hyper_rowset_chunk_t *, size_t *, size_t *, uint8_t * * *, size_t * *, int8_t * *)', got 'void(*)(struct hyper_rowset_chunk_t *, size_t *, size_t *, uint8_t * * *, size_t * *)'

pantab is incompatible with the version of Tableau Hyper API installed on your
system. Please upgrade both `tableauhyperapi` and `pantab` to the latest version.
If doing so does not fix this issue, please file an issue at
https://github.com/innobi/pantab/issues mentioning the exact pantab and HyperAPI
versions which triggered this error.

### PR DESCRIPTION
**Pull request by vogelsgesang**
The interface of the `hyper_rowset_chunk_field_values` function changed
in the latest version of HyperAPI. Update pantab to be compatible with
that HyperAPI version. While at it, also improve the error message to
include the exact version of HyperAPI

Fixes #145

### CODE REVIEW

**Comment by WillAyd** [comment]
@vogelsgesang this is huge thanks! We probably need to configure CI to have testing of both older versions and the latest version

**Comment by WillAyd** [comment]
For instance this may fix it for the latest version but then wouldn't work with older versions (at least I think from looking at it)

**Comment by vogelsgesang** [comment]
currently, this commit is written without backwards compatibility. Do we care about it enough to complicate the code?

The error message already states

> pantab is incompatible with version {} of Tableau Hyper API. **Please upgrade**
> **both `tableauhyperapi` and `pantab` to the latest version.** If doing so does not
> fix this issue, please file an issue at https://github.com/innobi/pantab/issues
> mentioning the exact pantab and HyperAPI versions which triggered this error.

Afaict, following that guidance should unblock all users

**Comment by WillAyd** [comment]
It's a little thorny. The problem is definitely going to be on the packaging side to enforce this without really muddying pip installs.

One last option is instead of dynamic runtime link to distribute the hyper lib / executable as part of pantab, which prevents the need for any runtime check. This is what I've done in HyperArrow as the tooling there improved a lot since this package was originally incepeted

https://github.com/innobi/hyperarrow/blob/main/.github/workflows/build_python_wheels.yml

**Comment by vogelsgesang** [comment]
<rant>
can we please pin `black` and `isort` to exact versions in the `environment.yml`? It's really annoying that I have to reformat many files which my commit didn't even touch, just because black & isort changed their formatting rules.
Pretty bad pull request experience. Thankfully, I know how to unblock myself, but this experience could easily demotivate other contributors...
</rant>

**Comment by WillAyd** [comment]
Yep also a great callout. I think using docker-compose and creating a lint rule will help a lot for that

**Comment by vogelsgesang** [comment]
> One last option is instead of dynamic runtime link to distribute the hyper lib / executable as part of pantab

That approach won't work for pantab given that pantab needs to use the exact same version of HyperAPI as also installed via pip. Otherwise, passing a Connection from Python to the C extension (e.g. to `frame_from_hyper_query`) would not work reliably. Afaict, the current architecture of HyperArrow won't allow for such "connection passing" functionality.

**Comment by vogelsgesang** [comment]
> It's a little thorny. The problem is definitely going to be on the packaging side to enforce this without really muddying pip installs.

Good point. I changed changed the `setup.py` to restrict the HyperAPI version

```
-    install_requires=["pandas", "tableauhyperapi", "numpy"],
+    install_requires=["pandas", "tableauhyperapi>=0.0.14567", "numpy"],
```

According to the [Python packaging user guide](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/), this should ensure that HyperAPI is recent enough. Not sure if it works on all package managers (conda, anaconda, ...), though. I have no experience with those package managers

**Comment by WillAyd** [comment]
That's a good point for re-used connections. One way to offer the same functionality (in a breaking manner) would be to have read / write function that can optionally return a connection. That way everything at the binary is totally managed within pantab and you don't even necessarily need tableauhyperapi or runtime linkage.

Hard to say what is the least evil going forward. Dynamic linkage over a C or C++ API that isn't stable can cause a lot of headaches with version compatibility that I think most Python users aren't going to really get

**Comment by vogelsgesang** [comment]
> Thankfully, I know how to unblock myself

Scratch that. It seems I am unable to get past
> ERROR: /home/runner/work/pantab/pantab/pantab/__init__.py Imports are incorrectly sorted and/or formatted.

I can't reproduce this issue locally when running `isort **/*.py -c` and have no glue what the CI is complaining about

**Comment by vogelsgesang** [comment]
Ok, I am giving up on the formatting issues...

Either disable the `isort` check temporarily and merge this commit as is or feel free to modify the commit however you see fit to fulfill your formatting requirements

**Comment by vogelsgesang** [comment]
also, you might want to consider to make formatting an independent step in your pipeline, such that the CI at least still provides feedback wether the functional test cases pass, even if the formatting is incorrect

**Comment by WillAyd** [comment]
> According to the [Python packaging user guide](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/), this should ensure that HyperAPI is recent enough. Not sure if it works on all package managers (conda, anaconda, ...), though. I have no experience with those package managers

Just for your awareness setup.py is used only by setuptools. For quite some time anaconda and setuptools were the two main distribution methods, with tools like flint and poetry amongst others having come around more recently. The growth of these tools and desires to clean up distribution / installation has yielded way to non-setup.py specifications like pyproject.toml, which was introduced in PEP 518 and extended to a bit in PEP 621 to cover what we might need here, like install dependencies

With that said, I know for a fact that pip and setuptools in coordination never had a real dependency resolver. Adding a version pin in setup.py for this project makes it so when you install this tool it will check your other dependencies, but never really after unless you try to reinstall the tool. Nothing with setup.py can truly resolve dependencies in a robust manner

conda goes about this in a different way by isolating an environment and having projects be much more explicit about their dependencies, while also adding in a real dependency solver that makes sure all conditions can be satisfied when installin into an environment. Unfortunately this project is not installable via conda as it requires in its current state the tableauhyperapi, which itself is not in conda. When declaring dependencies in conda your upstream dependencies also need to be available in that ecosystem, so that's unfortuantely not an option

We could do pyproject.toml and see if that helps for build systems, but I'm not if that ever really prevents this. Historically with binary dependencies you would either using conda (not an option) or build a wheel that includes a copy of those libraries. I think the latter of those two is still going to be our best bet

**Comment by WillAyd** [comment]
One other solution here I can think of is to make this a runtime check. We would want a compat object at both the Python and C layers and can branch in the code depending on version of the tableauhyperapi is being passed

**Comment by WillAyd** [comment]
If you want to rebase / merge on #147 should help with the linting. Broke out into a separate job as you suggested. Also you can locally now just run `docker compose run --rm lint` and it will do what needs to be done from a black / isort perspective

**Comment by WillAyd** [comment]
I've tried doing the compatability approach in #152 which compiles but fails at runtime. I am not sure if that will be a feasible approach but I've opened an SO article out there to see if someone with more expertise in C can help

https://stackoverflow.com/questions/71630729/provide-compatability-for-various-function-signatures-from-shared-library

File: pantab/__init__.py [hunk 1]
```diff
@@ -109,6 +113,6 @@ def _get_hapi_function(name, sig):
     ),
     _get_hapi_function(
         "hyper_rowset_chunk_field_values",
-        "struct hyper_error_t *(*)(struct hyper_rowset_chunk_t *, size_t *, size_t *, uint8_t * * *, size_t * *, int8_t * *)",
```
Comment (WillAyd) [inline pantab/__init__.py:112]: This is different than what is in tableauhyperapi.h - is that intentional?

Comment (WillAyd) [reply pantab/__init__.py:112]: Yea this was in reference to the `const` qualifiers - didn't realize cffi dropped those

**Comment by WillAyd** [comment]
Thanks @vogelsgesang

**Comment by vogelsgesang** [comment]
Thanks for pushing this over the finish line, @WillAyd 👍

**Pull request merged**